private/util.bzl

```diff
@@ -0,0 +1,38 @@
+load("@bazel_version//:version.bzl", "version")
+
+def _safe_int(s, v):
+  if not s.isdigit():
+    fail("invalid Bazel version '{}': non-numeric segment '{}'".format(v, s))
+  return int(s)
+
+def _partition(s):
+  for i in range(len(s)):
+    c = s[i]
+    if c == "-":
+      return s[:i], s[i+1:]
+    if not c.isdigit() and c != ".":
+      return s[:i], s[i:]
+  return s, ""
+
+def parse_version(v):
+  """Parses the given Bazel version string into a comparable value."""
+  if not v:
+    # An empty string is treated as a "dev version", which is greater than anything.
+    v = "999999.999999.999999"
+  release, prerelease = _partition(v.partition(" ")[0])
+  segments = release.split(".")
+  if len(segments) != 3:
+    fail("invalid Bazel version '{}': got {} dot-separated segments, want 3".format(v, len(segments)))
+  return [_safe_int(s, v) for s in segments], not prerelease, prerelease
+
+BAZEL_VERSION = parse_version(version)
+
+def le(v): return lambda: BAZEL_VERSION <= parse_version(v)
```

Comment:
I wonder whether we actually need this indirection - assuming this would only be evaluated once, couldn't we just do so eagerly? We could benchmark the overhead assuming something like < 1,000 features longterm.

Reply:
we probably don't need the indirection :) I was focused on the ease of defining a new feature like `some_feature = all_(ge("1.2.3"), le("2.3.4"))`, but `some_feature = lambda: ge("1.2.3") and le("2.3.4")` is just as flexible and even more readable.

Reply:
Assuming that the vast majority of features will be trivial (in terms of CPU time) to check for, we could also have the convention that struct members that read as field names (`is_supported`, `has_foo`) are plain, eagerly evaluated fields, and, if we even ever need them, have these lazily evaluated "accessors" be named like `check_for_...`, indicating that they are methods.

`if bazel_features.toolchains.has_optional_toolchains:` just feels more Pythonic/Starlarky and I could see forgetting the `()` being a relatively common source of errors/typos.

Reply:
ok, I'm convinced :)